srk/src/chc.ml

```diff
@@ -244,120 +492,119 @@ module ChcSrkZ3 = struct
-             BatHashtbl.find ind_to_sym index
-           else (
-             let sort = typ_of_sort (Z3.Expr.get_sort arg) in
-             let sym = 
-               mk_symbol srk ~name:(Z3.Symbol.to_string names.(index)) sort 
-             in
-             BatHashtbl.add ind_to_sym index sym;
-             sym)) 
+        (fun arg ->
+           begin match Z3.AST.get_ast_kind (Z3.Expr.ast_of_expr arg) with
+             | VAR_AST ->
+               let index = Z3.Quantifier.get_index arg in
+               BatHashtbl.find ind_to_sym index
+               (* Logic for handling case if rel atom arg is a numerical
+                * or logical constant *)
+             | NUMERAL_AST ->
+               let sym = mk_symbol srk `TyInt in
+               let q = SrkZ3.term_of_z3 srk arg in
+               eqs := (mk_eq srk (mk_const srk sym) q) :: !eqs;
+               sym
+             | APP_AST -> 
+               let decl = Z3.Expr.get_func_decl arg in
+               begin match Z3.FuncDecl.get_decl_kind decl with
+               | OP_TRUE ->
+                 let sym = mk_symbol srk `TyBool in
+                 eqs := (mk_const srk sym) :: !eqs;
+                 sym
+               | OP_FALSE -> 
+                 let sym = mk_symbol srk `TyBool in
+                 eqs := (mk_not srk (mk_const srk sym)) :: !eqs;
+                 sym
+               | _ -> invalid_arg "Cannot have non-trivial terms as param to relation"
+               end
+             | _ -> assert false
+           end) 
         (Z3.Expr.get_args z3pred) in
     let decl = Z3.Expr.get_func_decl z3pred in
     let rsym = Z3.Symbol.to_string (Z3.FuncDecl.get_name decl) in
     let relation = 
       if Hashtbl.mem rsym_to_int rsym then Hashtbl.find rsym_to_int rsym
       else (
-        let typ = List.map (fun arg -> typ_symbol srk arg) args in
+        let typ = List.map (fun arg -> typ_symbol_fo srk arg) args in
         let res = mk_relation fp ~name:rsym typ in
         Hashtbl.add rsym_to_int rsym res;
         res)
     in
-    relation, args
-
-  (* Similiar to above but always uses creates a fresh symbol. [eq_syms] tracks
-   * which fresh symbols we created for indices that already exist in 
-   * [ind_to_sym] *)
-  let rel_atom_of_z3_fresh srk fp ind_to_sym rsym_to_int names z3pred =
-    let fresh_index_map = BatHashtbl.create 91 in
-    let eq_syms = ref [] in
-    let atom = 
-      rel_atom_of_z3 srk fp fresh_index_map rsym_to_int names z3pred 
-    in
-    BatHashtbl.iter (fun ind sym ->
-        if BatHashtbl.mem ind_to_sym ind 
-        then eq_syms := ((BatHashtbl.find ind_to_sym ind), sym) :: !eq_syms
-        else BatHashtbl.add ind_to_sym ind sym)
-      fresh_index_map;
-    atom, !eq_syms
-
+    (relation, args), !eqs
 
   let parse_z3fp ?(z3queries=[]) srk fp z3fp =
     let rsym_to_int = BatHashtbl.create 91 in
     let decl_kind e = Z3.FuncDecl.get_decl_kind (Z3.Expr.get_func_decl e) in
     let parse_rule rule =
```

Comment:
Can you use SrkZ3.formula_of_z3 and then manipulate the srk formula?

Reply:
As per previous note - yes. I believe that should be fine.